IPC: IPM: Add ipm adsp driver & sanitycheck

finikorg opened the pull request:
Add IPM ADSP driver, DT configuration, ipm_echo sample, IPM console and documentation.
Added sanitycheck support for device-testing of up_squared_adsp.

Testing:
Build sample with
```
$ west build -p -b up_squared_adsp zephyr/samples/hello_world/
```
Flash with command
```
$ west flash zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh <!KEY>
```
Open simplest polling mailbox terminal with
```
$ sudo boards/xtensa/up_squared_adsp/tools/mbterm.py 
*** Booting Zephyr OS build zephyr-v2.2.0-1620-g4484b8eff384  ***
Hello World! up_squared_adsp
```
Try sanitycheck as superuser with

```
# sanitycheck -M -c -p up_squared_adsp --device-test \
   --west-flash="zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh, otc_private_key.pem" \
   --device-serial-pty zephyr/boards/xtensa/up_squared_adsp/tools/mbterm.py
```
Running as regular user with ambient (more about ambient: https://github.com/zephyrproject-rtos/zephyr/pull/22724#issuecomment-627308907)

```
$ ~/work/ambient/set_ambient zephyr/scripts/sanitycheck -M -c -p up_squared_adsp --device-test \
  --west-flash="zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh, otc_private_key.pem" \
  --device-serial-pty zephyr/boards/xtensa/up_squared_adsp/tools/mbterm.py -v
```

finikorg committed:
console: ipm_console: Add console over IPM

Add console over Inter Processor Mailboxes (IPM).
This is useful for AMP processors like ADSP found on up_squared board.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
ipm: ipm_adsp: Add ADSP IPM driver

Add IPM driver for ADSP.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: ipm_echo: Add sample and documentation

Add IPM echo sample echoing data over IPM. Add documentation for
the sample.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: intel_apl_adsp: Use IPM for firmware IPC

Use for IPC with firmware loader IPM API.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

dcpleung commented:
Could you squash the two sample commits together?

finikorg commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+/* FIXME define in DTS */
+#define DT_IPM_ADSP_0_BASE_ADDRESS	0x10
+#define DT_IPM_ADSP_0_NAME		"IPM_0"
+#define DT_IPM_ADSP_0_IRQ		SOC_AGGREGATE_IRQ(7, 6)
```
@dcpleung Do you think we need to put theses parameters to Device Tree? How can I use `SOC_AGGREGATE_IRQ(7, 6)` there?

dcpleung replied:
You can looking the dtsi file for `intel_s1000_crb`. The DMA controllers there are under the CAVS interrupt contoller.

finikorg replied:
Added bindings 3cafd24 and changed dtsi 6bf8fda.

finikorg committed:
dts: Add intel ADSP mailbox bindings

Add bindings for ADSP mailbox.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: intel_apl_adsp: Add mailbox to the board

Add mailbox to the board DTS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

andyross commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+	/* IPC Target Busy Interrupt Enable (IPCTBIE) */
+	if (dipct & IPC_DIPCT_BUSY && dipcctl & IPC_DIPCCTL_IPCTBIE) {
+		/* Mask Busy interrupt */
+		ipc_write(IPC_DIPCCTL, dipcctl & ~IPC_DIPCCTL_IPCTBIE);
```
So... you're in the ISR here already.  Why is this masking the interrupt it's servicing?  I'm guessing that this is either needless, or what is happening here is you're masking delivery of the interrupt to the OTHER side of the IPM channel while the shared state is indeterminate, in which case you really need a comment here (and probably need some other locking I'm guessing?).  Or maybe the mask/unmask sequence is needed to properly ACK the interrupt?  Same thing, needs a comment.

finikorg replied:
I do not have good documentation for this, only register description. I have followed what SOF is doing for IPC IRQ handler:
https://github.com/thesofproject/sof/blob/3dd843a46defd5234a517991a7c0049c66233b1d/src/drivers/intel/cavs/ipc.c#L64-L137
I will add more comments.

lgirdwood replied:
If I cast my mind back, we cant clear this IRQ until we have finished processing the IPC (which may take some time) hence the mask.  Clearing could also interrupt the host on some HW too. So short term this flow is fine as it survives 25k flood IPC pings. Long term we should be operating a msg queue where we can clear the IRQ and process the IPC offline (and allow other IPC to be sent).

andyross commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+	if (dipcie & IPC_DIPCIE_DONE && dipcctl & IPC_DIPCCTL_IPCIDIE) {
+		/* Mask Done interrupt */
+		ipc_write(IPC_DIPCCTL,
+			  ipc_read(IPC_DIPCCTL) & ~IPC_DIPCCTL_IPCIDIE);
```
Same comment as above.  Need clarity on what is being masked and why.

andyross commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+	}
+
+	memcpy((void *)IPM_ADSP_MAILBOX_OUT, data, size);
+	SOC_DCACHE_FLUSH((void *)IPM_ADSP_MAILBOX_OUT, size);
```
Given that this is an app-visible API, you almost certainly want to check size vs. MAILBOX_OUT_SIZE to avoid scribbling over shared areas and potentially confusing/crashing the host kernel.  Or at the very least check it in an __ASSERT().

So this supports only one message at a time?  A quick grep says that the outbox region is 4k, I think, so it seems like some queuing ability would really improve things.

finikorg replied:
About queuing. API mentions:
https://github.com/zephyrproject-rtos/zephyr/blob/bd5048c251bd578776d063ac45a94f8b482c5ed4/include/drivers/ipm.h#L118

finikorg replied:
> Or at the very least check it in an __ASSERT().

Added `BUILD_ASSERT`.

andyross commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+			/* Use zero copy */
+			data->callback(data->callback_ctx,
+				       dipct & IPC_DIPCI_MSG_MASK,
+				       (void *)IPM_ADSP_MAILBOX_IN);
```
Similar comment to the send side: this supports only one message in the buffer at a time.  But this buffer is 8k, and IPM is designed to be asynchronous, so some queing ability would definitely be possible and a good idea.

andyross reviewed (commented):
Some notes on first reading.  Is there a host-side driver integration somewhere?

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -0,0 +1,71 @@
+static int console_out(int c)
+{
+	/* Send character with busy wait */
+	ipm_send(ipm_console_device, 1, c & ~BIT(31), NULL, 0);
```
OK, excellent example of why you want queuing in this driver.  This is going to stall waiting on the other side to take an interrupt and clear the busy flat for... every... byte... of... log... output...  Some linux driver developer is going to be coming at you with a shiv made from the shards of a smartphone screen.

FWIW: accumulating a local line buffer and flushing when it gets full or on newline would be a very reasonable workaround too.

finikorg replied:
I took it initially from `drivers/console/ipm_console_sender.c` https://github.com/zephyrproject-rtos/zephyr/blob/49bf8b1d544fa932f1d012d2c9252fdf1021b167/drivers/console/ipm_console_sender.c#L28

I think we can just send without busy wait. If we get BUSY we can just drop the character.

For the testing it was easier to set this flag to verify that it works.

finikorg replied:
I have now immediate log with line buffer and another option using message queue msgq.

finikorg commented:
> Some notes on first reading. Is there a host-side driver integration somewhere?

For the host side I use devmem2, in this patch set there is README
https://github.com/finikorg/zephyr/tree/ipm/samples/subsys/ipc/ipm_echo

lgirdwood commented on soc/xtensa/intel_apl_adsp/adsp.c:
```diff
@@ -78,6 +78,14 @@ static void prepare_host_windows(void)
 	SOC_DCACHE_FLUSH((void *)(HP_SRAM_WIN0_BASE + SRAM_REG_FW_END),
 			 HP_SRAM_WIN0_SIZE - SRAM_REG_FW_END);
 
+	if (IS_ENABLED(CONFIG_IPM_ADSP)) {
```
I'm thinking long term we should build the extended FW boot data (i.e. data we send to host about window mappings and some FW capabilities via DT). Any thoughts on this, not something we need immediately but worth considering ?

finikorg replied:
At the moment we send window mappings:
https://github.com/zephyrproject-rtos/zephyr/blob/93b78b7bf8bfad75196de969a7487f95f235aad3/soc/xtensa/intel_apl_adsp/adsp.c#L91-107
On the host side we use https://github.com/thesofproject/sof-diagnostic-driver which does not use it much.

lgirdwood replied:
Linux and the diagnostic driver read this data as part of their IPC init, since mailbox sizes, access etc and IPC HW varies between the SoCs so I dont think there is anything stopping the DT upgrade when time permits.

lgirdwood reviewed (changes requested):
I think in general we need to look at de-duplicating code in APL and S1000 ADSP to create a SoC family for all CAVS IP versions.

finikorg commented on drivers/ipm/ipm_adsp.c:
```diff
@@ -0,0 +1,212 @@
+}
+
+static int ipm_adsp_send(struct device *dev, int wait, u32_t id,
+			 const void *data, int size)
```
At the moment I have access only to up_squared board, so it is the only one supported platform. I can test it also on other platforms when they become available.
BTW: How the code would be different for other SOCs?

lgirdwood replied:
Best to have a look at the IPC drivers in SOF, but generally the IPC doorbell HW has changed between 1st gen (BYT, CHT), 2nd gen (HSW, BDW), 3rd ((SKL - APL), 4th (CNL+). I've started some work on this now, will send a patch for you to review hopefully tomorrow.

finikorg replied:
I see the difference with `CAVS_VERSION_1_5` and others in https://github.com/thesofproject/sof/blob/master/src/drivers/intel/cavs/ipc.c
I can add support for other boards but I have only `up_squared_adsp`.

finikorg committed:
CODEOWNERS: Add ipm_adsp driver and ipm_console codeowner.

Add @finikorg as codeowner for ipm_adsp driver and ipm_console.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
soc: intel_apl_adsp: Add defaults for IPM and IPM console

Add defaults for IPM_ADSP and IPM_CONSOLE when IPM and CONSOLE are
enabled.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: up_squared_adsp: Add IPM and console to the board

Add by default IPM console for up_squared_adsp board, used for sanity
checks.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: logger: Filter out sanitycheck for ring buffer backend

For CONFIG_LOG_BACKEND_RB logs are printed to the ring buffer in memory,
so special tools are needed to read those logs.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
console: ipm_console: Add message queue to console

Add message queue to console in CONFIG_IPM_CONSOLE_IMMEDIATE is not
selected.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: up_squared_adsp: Add IPM terminal tool

Add tool which can access ADSP console over IPM using Python device
library and polling Doorbell registers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
console: ipm_console: Use line buffer before sending

Use line buffer to store characters before sending over IPM.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: up_squared_adsp: update terminal tool

Update terminal tool to use line buffer read instead of one charcter
over IPM.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

xuhao8210 commented:
After I tried this patch on the up_squared_adsp platform, I found that the sanitycheck script is be required root access or sudo to run as the script: ”mbterm.py” is be required root access . So could you please support to run sanitycheck via normal user ?  Thanks.

finikorg committed:
boards: up_squared_adsp: Update mailbox terminal

Update permissions for mmap.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

galak commented:
There's a lot in here that seems like it should get pulled into different PRs.
* new driver
* ipm console
* sanitycheck updates
...

finikorg commented:
> After I tried this patch on the up_squared_adsp platform, I found that the sanitycheck script is be required root access or sudo to run as the script: ”mbterm.py” is be required root access . So could you please support to run sanitycheck via normal user ? Thanks.

There is a solution running as a normal user using ambient caps (assuming /dev/hda is easily managed since it is diagnostic driver):

`$ git clone https://github.com/finikorg/ambient.git`

Get libcap-ng-dev

`$ sudo apt install libcap-ng-dev`

Run make to build set_ambient
Set caps to the executable with

` $ sudo setcap cap_sys_rawio+ep set_ambient`

Run mbterm with ambient caps
```
$ ~/work/ambient/set_ambient boards/xtensa/up_squared_adsp/tools/mbterm.py 
Starting process with CAP_SYS_RAWIO in ambient
```
In another terminal run flash

`$ west flash zephyr/boards/xtensa/up_squared_adsp/tools/up_squared_adsp_flash.sh otc_private_key.pem`

Now you can see output from mbterm:
```
*** Booting Zephyr OS build zephyr-v2.2.0-1620-g4484b8eff384 ***
Hello World! up_squared_adsp
```

andyross commented on dts/xtensa/intel/intel_apl_adsp.dtsi:
```diff
@@ -92,6 +92,14 @@
+			reg = <0x1180 0x20>;
+			interrupts = <0x7 0 3>;
+			interrupt-parent = <&cavs0>;
+			label = "IPM_0";
```
This is a circumstance where IMHO devicetree is inappropriate.  The mailbox is a 100% pure software construction.  I think this wants to go in kconfig.

finikorg replied:
@andyross, I think I am following like it is done for idc (although it was added after my PR)
https://github.com/zephyrproject-rtos/zephyr/blob/52992b065829d8280119fca96ab27d49e9d4b5fc/dts/xtensa/intel/intel_apl_adsp.dtsi#L89-L94
I have to say that this simplifies describing IRQ number.

andyross replied:
OK, maybe I'm just confused then.  I thought this was implemented on top of the IDC mechanism.  Is it a separate hardware device?  If so, we should probably have a driver for that, which can go into DT, right?  This seems all tied up with the IPM driver, which is the software protocol.

andrewboie replied:
IPM driver subsystem is for mailbox-type CPU features. The first use-case was the mailboxes that the x86 and ARC CPUs used to talk to each other on Arduino 101. I wouldn't go as far to say that it's purely a software construction.

It got named IPM for "inter processor mailbox" as the generic term "mailbox" is already used by a generic kernel object type.

andrewboie replied:
I might be missing something though, are we instantiating IPM on this HW without an actual HW mechanism for communicating to other CPUs?

andyross replied:
There's definitely a mechanism.  There is an "IDC" set of registers for delivering interrupts to other cores (both the host CPU and the DSP cores), and there's a shared mapping into the DSP SRAM that exists on both sides of the link.  This part works.  My confusion is that I thought the IPM was on top of this, but the DTS block above implies that it's different hardware.  I **think** that's just an alias of the existing IDC registers/IRQ, which would be wrong.  But the DSP is a big place and I'm still learning the map. :)

dcpleung replied:
I believe this is using the IPC mechanism which is used for host-to-DSP communication. IDC is for DSP-to-DSP.

finikorg replied:
@andyross, @dcpleung explained right the difference between those IPC. In a way they should be described in the similar way in DTS.

andyross commented on samples/subsys/logging/logger/sample.yaml:
```diff
@@ -5,6 +5,7 @@ sample:
 tests:
   sample.logger.basic:
     tags: logging
+    filter: not CONFIG_LOG_BACKEND_RB
```
Can't we integrate the tooling with sanitycheck instead?  Whitelisting tests (OK, this is technicaly a sample) is a last resort to work around bugs or platform misfeatures that have no reasonable resolution.  There's no reason we can't add the host-side log dumper to the sanitycheck console integrations.

andyross commented on dts/bindings/ipm/intel,intel-adsp-mailbox.yaml:
```diff
@@ -0,0 +1,12 @@
+
+description: INTEL ADSP MAILBOX
+
+compatible: "intel,intel-adsp-mailbox"
```
Same complaint as elsewhere -- I'm not super comfortable seeing software protocols end up in devicetree configuration.

finikorg replied:
Already replied above.

andyross commented on soc/xtensa/intel_apl_adsp/Kconfig.defconfig:
```diff
@@ -44,7 +44,7 @@ config TEST_LOGGING_DEFAULTS
 
 config LOG_PRINTK
-	default y
+	default y if !CONSOLE
```
This changes the default for everyone everywhere, that sounds wrong.

andyross commented on drivers/console/Kconfig:
```diff
@@ -190,6 +190,30 @@ config IPM_CONSOLE_ON_DEV_NAME
+
+config IPM_CONSOLE_QUEUE_SIZE
+	int "IPM console queue size"
+	default 4096
```
That is a VERY large stack given what this thing is doing, even for Xtensa code.

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -11,12 +11,62 @@
+				character, ret);
+		}
+
+		k_yield();
```
Why?  This loop is going to jump straight into k_msgq_get() and block.  Please remove this yield, or if it's really needed let's figure out the root cause of the hack (likely a thread priority messup) and fix that instead.

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -11,12 +11,62 @@
+#else
+	char character = c;
+
+	ret = k_msgq_put(&msg_queue, &character, K_NO_WAIT);
```
What's the design goal here?  Why are we implementing asynchronous console output and for who?  The older design was a simple ring buffer which was async by design without the need for complicated and expensive IPC on the device.

Basically this driver creeps me out a bit.  What's the point?

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -46,26 +96,38 @@ static void ipm_console_hook_install(void)
+	k_thread_create(&thread_data, thread_stack,
+			K_THREAD_STACK_SIZEOF(thread_stack),
+			msgq_thread, ipm, NULL, NULL,
+			CONFIG_MAIN_THREAD_PRIORITY - 1,
```
If this driver is carrying console output (!), I can pretty much guarantee that "one higher than the app main thread priority" is the wrong value to hard code.

I mean.... what happens when someone spins in a high priority thread and blocks console output?  The system deadlocks silently, right?

andyross commented on drivers/console/Kconfig:
```diff
@@ -207,7 +215,7 @@ config IPM_CONSOLE_THREAD
 config IPM_CONSOLE_QUEUE_SIZE
 	int "IPM console queue size"
-	default 4096
+	default 16
```
That seems like a pretty impactful size change!  Is this moving from characters to lines or something?  Seems like it needs documentation about what the units are here.

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -52,22 +53,30 @@ static void msgq_thread(void *arg1, void *arg2, void *arg3)
+
 #if defined(CONFIG_IPM_CONSOLE_IMMEDIATE)
-	ret = ipm_send(ipm_dev, 1, c & ~BIT(31), NULL, 0);
+	ret = ipm_send(ipm_dev, 1, len & ~BIT(31), buf, len);
```
I'm seeing this pattern a lot in this pull request, where you're setting the high bit of the IPM message ID.   Now, that's fine as far as IPM itself goes -- the ID is just an opaque value.  But clearly it means something, and I can't see what.

What's going on here?  This needs documentation, I think.  Right now the code you're adding is speaking an "ADSP IPM" variant protocol and won't work with default IPM backends.  That's not crazy, since they're all device-specific, but it needs some kind of framework in software I think and not just a bunch of raw bits flying around.

finikorg replied:
Here I have changed, following your suggestion, from sending one character to sending a line. This in not IPM but user-defined "console over IPM", kind of higher level protocol.

andyross replied:
OK, then maybe a "IPM_CONSOLE_MSGID(len)" macro would make this clearer?

andyross replied:
Oh, and note that this presumes a 32-bit-clean IPM message ID.  Note that IPM implementations are allowed to have a smaller ID space, there's an API to query it.  This should probably be matched with an __ASSERT (or BUILD_ASSERT) to make sure no one tries this with an incompatibile backend.

andyross replied:
OK, I've figured out what you're doing.  You're clearing this bit because the hardware reserves the top bit of the message register for the BUSY flag and you don't want to clobber anything.  But the API already gives you a way to inform users of that: `ipm_max_id_val_get()`.  So implement that API and put this bit clear down into the driver level underneath `ipm_send()`.

andyross commented on drivers/console/ipm_console.c:
```diff
@@ -52,22 +53,30 @@ static void msgq_thread(void *arg1, void *arg2, void *arg3)
 static int console_out(int c)
 {
+	static char buf[CONFIG_IPM_CONSOLE_LINE_BUF_LEN];
+	static size_t len;
```
Where is the locking for this data?  It something happening up the stack somewhere that makes this safe?  Remember that console output can happen from interrupt handlers and across SMP cores, and by default this code isn't safe.  I think you need a spinlock around this data.  If you don't need a spinlock, you need a comment explaining why.

andyross commented on boards/xtensa/up_squared_adsp/tools/mbterm.py:
```diff
@@ -14,8 +14,8 @@
-with open("/dev/mem", "r+b") as f:
-    mem_map = mmap(f.fileno(), LENGTH, offset=MBOX)
+with open("/dev/mem", "rb") as f:
+    mem_map = mmap(f.fileno(), LENGTH, access=ACCESS_COPY, offset=MBOX)
```
I had to look this up (because **of course** Python doesn't expose the system's mmap syscall with the same API). 

"ACCESS_COPY" specifies a copy-on-write mapping.  That makes basically zero sense to me for a shared region.  Why on earth would this python script want its own copy of the memory that the other side was guaranteed not to see?  That's not a "permission", and I don't understand why you'd want it here at all.  What's going on?

finikorg replied:
I have probably screwed something with this, should be `ACCESS_READ`, will fix.

finikorg replied:
@andyross I cherry-picked this patch to #27804, addressed most of the comments except for this one, it needs some modification since from_buffer does not work with ACCESS_READ. I'll try to find some solution for this.

andyross reviewed (changes requested):
Some feedback.  Mostly nitpicks or easily resolved issues.

But I will say... I kinda hate that async console handler.  I just don't see how that's going to do anything but create bugs and hardship.  What's the value being provided?  Note that the logging subsystem already has an async mode for applications that require that it not impact latency or performance (FWIW: I kinda hate that too, but I see why it exists).  Why can't we just use that?

Also, that COW thing in the python script has me completely confused and really needs to be explained.

nashif commented:
@finikorg I would like to see the sanitycheck patch split out of this PR and submitted individually. 
I still do not understand why you need special handling for pty devices and why can't we access this as just any other serial device.  so split this into another PR to have the discussion there.

andyross commented:
Yeah, I found the I**P**C vs. I**D**C distinction in the docs (I guess I had filed them under the same name in my head -- the mechanisms are broadly similar AFAICT, and it seems like both of them have IPM backends). 

If I can ask for a light rearchitecture here: what you really want is to split the driver into separate layers for SoC devices and Zephyr layers.  You want a single "IPC" driver (or just SoC API, no need to make it a struct device given how low-level this is) to manage host/DSP interrupt delivery, a single[1] "IDC" driver for DSP/DSP interrupt delivery, it should specify its registers and IRQs in devicetree, and it should do nothing else.

Then underneath that you can layer hooks for subsystems that require them, like IPM/console and arch_sched_ipi.  It's perfectly fine for these subsystem abstractions to be zero cost (i.e. just call out to IPM inside an ifdef), but they should be separate code for understanding, compartmentalization and review.

[1] Or maybe combine these two into a single API at the SoC layer if the register layouts really are compatible like I think they might be.

andyross commented:
Actually reading more carefully: the mechanisms are exactly the same.  The host/dsp, csme/dsp and dsp/dsp interrupt targetting all share the same 4-register interface and work the same way[1].  We really should be able to have a tiny SoC driver that implements support for all of these symmetrically.

[1] With the expected few tiny glitches like where the register with the interrupt enable bits is, and whether or not hardware propagates the BUSY flag from the target into the high bits of the "extension" word at the initiator register.

finikorg commented:
@andyross when I send PR there were no IDC, otherwise I would probably try to make it general.

finikorg commented:
> Some feedback. Mostly nitpicks or easily resolved issues.
> 
> But I will say... I kinda hate that async console handler. I just don't see how that's going to do anything but create bugs and hardship. What's the value being provided? Note that the logging subsystem already has an async mode for applications that require that it not impact latency or performance (FWIW: I kinda hate that too, but I see why it exists). Why can't we just use that?
> 
> Also, that COW thing in the python script has me completely confused and really needs to be explained.

Logging subsystem for ADSP works through shared memory ring buffer, on the host we can access shared memory ring buffer through /dev/mem. With mailboxes we poll Doorbell register and if there is a new message read it. Though polling Doorbell is not optimal but still better then reading /dev/mem all the time and somehow compare do we have new line there.

finikorg commented:
@andyross host/dsp and csme/dsp IPC can be probably generalized, but for dsp-dsp I do not see a way to make it more general with less code. There is also some processing of ID here:
https://github.com/finikorg/zephyr/blob/54de9001be2127586cab958c168c7c091cce5a85/drivers/ipm/ipm_cavs_idc.c#L60
and another issue is that I use shared memory to pass message, IDC uses Message Extension (MSGEXT) so it is not able to send data:
https://github.com/finikorg/zephyr/blob/54de9001be2127586cab958c168c7c091cce5a85/drivers/ipm/ipm_cavs_idc.c#L141-L149

dcpleung commented:
??? I thought the driver has already been merged.

finikorg commented:
> ??? I thought the driver has already been merged.

I have split this PR to small chunks, IPM driver was merged, remaining are console over IPM and tools.

finikorg commented:
The PR was split to smaller logical entities and merged upstream. Closing this.
